test/functional/finalization_deposit_reorg.py

```diff
@@ -0,0 +1,218 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019 The Unit-e developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+from test_framework.test_framework import UnitETestFramework
+from test_framework.util import (
+    connect_nodes,
+    sync_blocks,
+    assert_equal,
+    disconnect_nodes,
+    wait_until,
+)
+import os
+
+ESPERANZA_CONFIG = '-esperanzaconfig={"epochLength": 10}'
+
+
+class EsperanzaDepositReorgTest(UnitETestFramework):
+    def set_test_params(self):
+        self.num_nodes = 4
+
+        self.extra_args = [
+            [ESPERANZA_CONFIG],
+            [ESPERANZA_CONFIG],
+            [ESPERANZA_CONFIG],
+            [ESPERANZA_CONFIG, '-validating=1'],
+        ]
+        self.setup_clean_chain = True
+
+    def setup_network(self):
+        # topology:
+        # node[0] -> node[1]
+        #         -> node[2]
+        #         -> node[3]
+        self.setup_nodes()
+        connect_nodes(self.nodes[0], 1)
+        connect_nodes(self.nodes[0], 2)
+        connect_nodes(self.nodes[0], 3)
+
+    def run_test(self):
+        node0 = self.nodes[0]
+        node1 = self.nodes[1]
+        node2 = self.nodes[2]
+        finalizer = self.nodes[3]
+
+        self.setup_stake_coins(node0, node1)
+
+        # leave IBD
+        # e0 - e1[1] node1, node2, node3, finalizer
```

Comment:
At the moment, `self.generate_sync` is most convenient way to leave IBD.

Reply:
It does a little bit too much, e.g, syncs `mempools` and `validationinterfacequeue`. 
As I recall it was designed to tackle the issue "vote in the future". 

@AM5800 do you think we need to this function at all as currently, we replaced it with `wait_for_vote_and_disconnect`?. Can we mark it as deprecated and eventually delete it?

Reply:
`generate_sync` was designed to handle the fact that block generation might be faster than vote casting(since they are on different threads). But it turned out that this is not enough to achieve consistent state among nodes.

So yes, it can be deprecated.

Reply:
I am resolving it.